File: mvx/mvc/spring-mvc/src/main/java/com/microwind/springbootorder/services/order/OrderService.java
package com.microwind.springbootorder.services.order;

import com.microwind.springbootorder.models.order.Order;
import com.microwind.springbootorder.repository.order.OrderRepository;
import com.microwind.springbootorder.utils.NotFoundException;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class OrderService {
  private final OrderRepository orderRepository;
  private final OrderMapper orderMapper;

  // 创建订单
  public Order createOrder(Order order) {
    // 业务逻辑，例如生成订单编号等
    if (order.getOrderNo() == null || order.getOrderNo().isEmpty()) {
      String orderNo = "ORD-" + UUID.randomUUID().toString().replace("-", "").substring(0, 16);
      order.setOrderNo(orderNo);
    }
    if (order.getStatus() == null) {
      order.setStatus(Order.OrderStatus.CREATED);
    }
    return orderRepository.save(order);
  }

  // 根据订单编号查询订单
  public Order getByOrderNo(String orderNo) {
    return orderRepository.findByOrderNo(orderNo)
            .orElseThrow(() -> new NotFoundException("Order not found with orderNo: " + orderNo));
  }

  // 查询用户订单列表
  public List<Order> getUserOrders(Long userId) {
    return orderRepository.findByUserId(userId);
  }

  // 更新订单状态
  @Transactional
  public Order updateOrderStatus(String orderNo, Order.OrderStatus status) {
    // 1. 先取后存
    /*
     Order order = getByOrderNo(orderNo);
     order.setStatus(status);
     return orderRepository.save(order);
     */

    // 2. 更新状态后再查询，需要注意事务的缓存问题
    int updatedRows = orderRepository.updateOrderStatus(orderNo, status);
    if (updatedRows > 0) {
      return getByOrderNo(orderNo);
    }
    throw new EntityNotFoundException("Order update failed, order not found: " + orderNo);
  }


  // 更新订单
  public Order updateOrder(String orderNo, Order order) {
    Order existingOrder = orderRepository.findByOrderNo(orderNo)
            .orElseThrow(() -> new NotFoundException("Order not found with orderNo: " + orderNo));
    if (order.getUserId()!= null) {
      existingOrder.setUserId(order.getUserId());
    }
    if (order.getAmount()!= null) {
      existingOrder.setAmount(order.getAmount());
    }
    if (order.getStatus()!= null) {
      existingOrder.setStatus(order.getStatus());
    }
    if (order.getOrderName()!= null) {
      existingOrder.setOrderName(order.getOrderName());
    }
    return orderRepository.save(existingOrder);
  }

  // 删除订单
  public void deleteOrder(String orderNo) {
    Order order = orderRepository.findByOrderNo(orderNo)
            .orElseThrow(() -> new NotFoundException("Order not found with orderNo: " + orderNo));
    orderRepository.delete(order);
  }

  // 获取所有订单
  public Page<Order> getAllOrders(Pageable pageable) {
    // 调用默认JPA方法，简单
//    return orderRepository.findAll(pageable);
    // 调用自定义方法，个性化
    return orderRepository.findAllOrders(pageable);
  }
}

File: mvx/mvc/gin-mvc/internal/services/order/order_service.go
package services

import (
  "errors"
  "fmt"
  models "gin-order/internal/models/order"
  repository "gin-order/internal/repository/order"
  "strings"

  "github.com/google/uuid"
  "github.com/shopspring/decimal"
)

// GenerateOrderNo 使用 UUID 生成订单编号
func GenerateOrderNo() string {
  return "ORD-" + strings.ReplaceAll(uuid.New().String(), "-", "")[:16]
}

// OrderService 定义订单服务接口[可选]
type OrderService interface {
  CreateOrder(o models.Order) (*models.Order, error)
  GetByOrderNo(orderNo string) (*models.Order, error)
  GetAllOrders(page, pageSize int) ([]models.Order, int, error)
  GetOrdersByUserID(userId string, page, pageSize int) ([]models.Order, int, error)
  UpdateOrder(o models.Order) (*models.Order, error)
  UpdateOrderStatus(orderNo, status string) (*models.Order, error)
  DeleteOrder(orderNo string) error
}

// OrderServiceImpl 订单服务实现 OrderService 接口
// 如不实现接口，也可以只定义struct，接口会更加利于扩展
type OrderServiceImpl struct {
  repo repository.OrderRepository
}

// NewOrderService 创建订单服务实例，并返回 OrderServiceInterface 接口
func NewOrderServiceImpl(repo repository.OrderRepository) OrderService {
  return &OrderServiceImpl{repo: repo}
}

// CreateOrder 创建订单
func (s *OrderServiceImpl) CreateOrder(o models.Order) (*models.Order, error) {
  zero := decimal.NewFromInt(0)
  o.OrderNo = GenerateOrderNo()
  if o.OrderName == "" || o.Amount.LessThanOrEqual(zero) || o.UserID == 0 {
    return nil, errors.New("无效的订单数据")
  }
  fmt.Println("CreateOrder:", o)
  return s.repo.CreateOrder(o)
}

// GetByOrderNo 通过订单编号获取订单
func (s *OrderServiceImpl) GetByOrderNo(orderNo string) (*models.Order, error) {
  return s.repo.GetByOrderNo(orderNo)
}

// GetAllOrders 获取所有订单（支持分页）
func (s *OrderServiceImpl) GetAllOrders(page, pageSize int) ([]models.Order, int, error) {
  orders, total, err := s.repo.GetAllOrders(page, pageSize)
  if err != nil {
    return nil, 0, fmt.Errorf("GetAllOrders service error: %w", err)
  }
  return orders, total, nil
}

// GetOrdersByUserID 获取用户的订单（支持分页）
func (s *OrderServiceImpl) GetOrdersByUserID(userId string, page, pageSize int) ([]models.Order, int, error) {
  orders, total, err := s.repo.GetOrdersByUserID(userId, page, pageSize)
  if err != nil {
    return nil, 0, fmt.Errorf("GetOrdersByUserID service error: %w", err)
  }
  return orders, total, nil
}

// UpdateOrder 更新订单
func (s *OrderServiceImpl) UpdateOrder(o models.Order) (*models.Order, error) {
  if o.OrderNo == "" {
    return nil, errors.New("订单ID是必需的")
  }
  return s.repo.UpdateOrder(o)
}

// UpdateOrderStatus 修改订单状态
func (s *OrderServiceImpl) UpdateOrderStatus(orderNo, status string) (*models.Order, error) {
  return s.repo.UpdateOrderStatus(orderNo, status)
}

// DeleteOrder 删除订单
func (s *OrderServiceImpl) DeleteOrder(orderNo string) error {
  return s.repo.DeleteOrder(orderNo)
}
